src/fsharp/ErrorResolutionHints.fs

```diff
@@ -23,60 +56,54 @@ let IsInEditDistanceProximity idText suggestion =
-        if not (name.StartsWithOrdinal("( ") && name.EndsWithOrdinal(" )")) then
-            false
-        else
-            let name = name.[2..name.Length - 3]
-            name |> Seq.forall (fun c -> c <> ' ')
+/// Filters suggestionss based on edit distance to the given unknown identifier.
+let FilterSuggestions (buffer: MaxBuffer<_,string>) (idText:string) =    
+    let uppercaseText = idText.ToUpperInvariant()
 
-    if allSuggestions.Contains idText then [] else // some other parsing error occurred
     let dotIdText = "." + idText
-    allSuggestions
-    |> Seq.choose (fun suggestion ->
+
+    fun (suggestion:string) ->
         // Because beginning a name with _ is used both to indicate an unused
         // value as well as to formally squelch the associated compiler
         // error/warning (FS1182), we remove such names from the suggestions,
         // both to prevent accidental usages as well as to encourage good taste
-        if IsOperatorName suggestion || suggestion.StartsWithOrdinal("_") then None else
-        let suggestion:string = demangle suggestion
-        let suggestedText = suggestion.ToUpperInvariant()
-        let similarity = EditDistance.JaroWinklerDistance uppercaseText suggestedText
-        if similarity >= highConfidenceThreshold || suggestion.EndsWithOrdinal(dotIdText) then
-            Some(similarity, suggestion)
-        elif similarity < minThresholdForSuggestions && suggestedText.Length > minStringLengthForThreshold then
-            None
-        elif IsInEditDistanceProximity uppercaseText suggestedText then
-            Some(similarity, suggestion)
+        if IsOperatorName suggestion || suggestion.StartsWithOrdinal("_") then 
+            ()
         else
-            None)
-    |> Seq.sortByDescending fst
-    |> Seq.mapi (fun i x -> i, x) 
-    |> Seq.takeWhile (fun (i, _) -> i < maxSuggestions) 
-    |> Seq.map snd 
-    |> Seq.toList
+            let suggestion:string = DemangleSuggestion suggestion
+            let suggestedText = suggestion.ToUpperInvariant()
+            let similarity = EditDistance.JaroWinklerDistance uppercaseText suggestedText
+            if similarity >= highConfidenceThreshold || suggestion.EndsWithOrdinal(dotIdText) then
+                buffer.Insert(similarity, suggestion) |> ignore
+            elif similarity < minThresholdForSuggestions && suggestedText.Length > minStringLengthForThreshold then
+                ()
+            elif IsInEditDistanceProximity uppercaseText suggestedText then
+                buffer.Insert(similarity, suggestion) |> ignore
 
-/// Formats the given predictions according to the error style.
-let FormatPredictions normalizeF (predictions: (float * string) list) =
-    match predictions with
-    | [] -> System.String.Empty
-    | _ ->
+/// Formats the given suggestions according to the error style.
+let FormatSuggestions normalizeF (idText:string) (buffer: ErrorLogger.MaxBuffer<_,string>) =
+    let values = buffer.Values()
```

Comment:
can we use IEnumerable interface on Maxbuffer like one I provided to reduce allocs etc?

Reply:
maybe.